PORT-9458 | Create automation that sends a Slack message for prs more than 3 days (#1447)
port-labs/port-docs: docs/actions-and-automations/define-automations/examples.md

---

## Send a Slack message when a PR is open for more than 3 days

> 

Reviewer: I'll prefer you add a one sentence overview section describing the purpose of this automation, before the prerequisite

### Prerequisites

To use this automation, ensure you have:

> - Port installed on GitHub.

Reviewer: ```suggestion
- Install Port's GitHub app by clicking [here](https://github.com/apps/getport-io/installations/new).
- Configure a [Slack app](https://api.slack.com/apps) that can post a message to a Slack channel. The app should have a `chat:write` bot scope under OAuth & Permissions.
```

### Integrate GitHub resources with Port

The goal of this section is to fill the software catalog with data directly from your GitHub organization. Port's GitHub app allows you to import `repositories`, `pull requests`, `workflows`, `teams` and other GitHub objects. For the purpose of this guide, we shall focus on pull requests (PR) object only. Follow the steps below to ingest your PR data to Port.

:::info Note
For the GitHub app installation, you will need to have a registered organization in Port and your Port user role must be set to `Admin`.
:::

1. **Install Port's GitHub app:**
   Follow the [installation guide](https://docs.getport.io/build-your-software-catalog/sync-data-to-catalog/git/github/installation).

> 2. **Check if `Pull Request` <PortTooltip id="blueprint">blueprint</PortTooltip> exists:**

Reviewer: ```suggestion
2. **Create the `Pull Request` blueprint (if it doesn't exist):**
```

>     - Go to your [Builder](https://app.getport.io/settings/data-model)
>     - If the `Pull Request` <PortTooltip id="blueprint">blueprint</PortTooltip> exists, proceed to step 3 to add the `openDuration` property.
>     - If the `Pull Request` <PortTooltip id="blueprint">blueprint</PortTooltip> does not exist, create it with the schema provided below.

Reviewer: ```suggestion
Go to your [Builder](https://app.getport.io/settings/data-model) page and create the `Pull Request` blueprint using the schema provided below
```

                "closed"
                ],
                "enumColors": {
                "merged": "purple",
                "open": "green",
                "closed": "red"
                }
            },
            "closedAt": {
                "title": "Closed At",
                "type": "string",
                "format": "date-time"
            },
            "updatedAt": {
                "title": "Updated At",
                "type": "string",
                "format": "date-time"
            },
            "mergedAt": {
                "title": "Merged At",
                "type": "string",
                "format": "date-time"
            },
            "link": {
                "type": "string",
                "format": "url"
            },
            "openDuration": {
                "title": "Open Duration",
                "type": "string",
                "format": "timer",
                "default": "2024-08-01T12:25:23.000Z"
            }
            },
            "required": []
        },
        "mirrorProperties": {},
        "calculationProperties": {},
        "aggregationProperties": {},
        "relations": {}
    }
    ```
    </details>

> 3. **Add `openDuration` **property** to `Pull Request` <PortTooltip id="blueprint">blueprint</PortTooltip> (if it already exists):**

Reviewer: ```suggestion
3. **Add an `openDuration` property to the `Pull Request` blueprint (if it already exists):**
```

>     - Navigate to the `Pull Request` <PortTooltip id="blueprint">blueprint</PortTooltip> in your Port Builder.

Reviewer: ```suggestion
    - Navigate to the `Pull Request` blueprint in your Port Builder.
```

            query: "true"
        port:
            entity:
            mappings:
                identifier: ".head.repo.name + '-' + (.number|tostring)" # The Entity identifier will be the repository name + the pull request number
                title: ".title"
                blueprint: '"githubPullRequest"'
                properties:
                    creator: ".user.login"
                    assignees: "[.assignees[].login]"
                    reviewers: "[.requested_reviewers[].login]"
                    status: ".state"
                    closedAt: ".closed_at"
                    updatedAt: ".updated_at"
                    mergedAt: ".merged_at"
                    prNumber: ".id"
                    link: ".html_url"
                    openDuration: "((now | tonumber) + (3 * 24 * 60 * 60) | todateiso8601)" # For 1-minute timer, use ((now | tonumber) + 60 | todateiso8601)
    ```

    </details>

   You should now be able to see your GitHub pull requests ingested successfully in the software catalog.

### Create a Slack Webhook

To send messages to Slack, you need to create a Slack webhook. Follow the instructions below to create one:

1. **Create a Slack App:**

>     - Go to [Slack API: Applications](https://api.slack.com/apps).

Reviewer: ```suggestion
    - Go to [Slack API](https://api.slack.com/apps).
```

    - Click "Create New App"
    - Select "From scratch" and provide an app name and the workspace where the app will be installed.

2. **Add Incoming Webhooks:**
    - In your app settings, go to "Incoming Webhooks"
    - Click "Activate Incoming Webhooks"

3. **Create a Webhook URL:**
    - Click "Add New Webhook to Workspace"
    - Select a channel where the messages will be posted and click "Allow"
    - Copy the generated webhook URL.

> 

Reviewer: I don't think we should spent much time showing users how to generate a Slack webhook URL. Let's just give them a [reference link from Slack](https://api.slack.com/messaging/webhooks) on how to create one

Author: okay

```json showLineNumbers
{
  "identifier": "prOpenForMoreThan3Days",
  "title": "Notify Slack on PR Open for More Than 3 Days",
  "icon": "Slack",
  "description": "Sends a Slack message when a PR has been open for more than 3 Days.",
  "trigger": {
    "type": "automation",
    "event": {
      "type": "TIMER_PROPERTY_EXPIRED",
      "blueprintIdentifier": "githubPullRequest",
      "propertyIdentifier": "openDuration"
    },
    "condition": {
      "type": "JQ",
      "expressions": [
        ".diff.after.properties.status == \"open\""
      ],
      "combinator": "and"
    }
  },
  "invocationMethod": {
    "type": "WEBHOOK",
    "url": "<Your Generated Slack Webhook>",
    "agent": false,
    "synchronized": true,
    "body": {
      "text": ":warning: *PR Overdue Notification*\n\n:page_facing_up: *Title:* {{ .event.diff.after.title }}\n\n:link: *Link:* <{{ .event.diff.after.properties.link }}|View PR>\n\n:bust_in_silhouette: *Creator:* {{ .event.diff.after.properties.creator }}\n\n:busts_in_silhouette: *Assignees:* {{ .event.diff.after.properties.assignees }}\n\n:eyes: *Reviewers:* {{ .event.diff.after.properties.reviewers }}"

> ## Update a service's default values upon creation

Reviewer: Please check this section on the AWS Amplifiy [link](https://pr-1447.d2ngvl90zqbob8.amplifyapp.com/actions-and-automations/define-automations/examples#automation-definition-3). It seem to have broken

}
```

## Run History

- To check your run history click on this icon on the top right conner beneath the `Builder` icon

<br></br>
<img src='/img/automations/automationRunHistoryPointer.png' width='60%' />
<br></br>

- Click on it to view your run history
<br></br>
<img src='/img/automations/automationRunHistory.png' width='60%' />

> <br></br>

Reviewer: I don't think this section will be needed sine we want to keep it short

Author: okay
